### PR TITLE
feat(csharp): adds a missed first or default opprtunity rule

### PR DESCRIPTION
**Pull request by baywet**
there are already rules for missed where/all/oftype/select/cast opportunities. It only makes sense to have FirstOrDefault which is a very common use case. Related #22484

### CODE REVIEW

File: csharp/ql/src/codeql-suites/csharp-security-and-quality.qls [hunk 1]
```diff
@@ -51,6 +51,7 @@
       - cs/linq/inconsistent-enumeration
       - cs/linq/missed-all
       - cs/linq/missed-cast
+      - cs/linq/missed-firstordefault
```
Comment (michaelnebel) [inline csharp/ql/src/codeql-suites/csharp-security-and-quality.qls]: We prefer not to add any new quality queries to the `csharp-security-and-quality` (the tags added to the query means that the query will be included in the `code-quality` suite).

File: csharp/ql/src/Linq/MissedFirstOrDefaultOpportunity.ql [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+import csharp
+import Linq.Helpers
+
+from ForeachStmtGenericEnumerable fes, IfStmt is, ReturnStmt ret, ReturnStmt defaultRet
+where missedFirstOrDefaultOpportunity(fes, is, ret, defaultRet)
+select fes,
+  "This foreach loop $@ - consider finding the element explicitly using '.FirstOrDefault(...)'.",
+  is.getCondition(), "returns the first sequence element satisfying a predicate"
```
Comment (michaelnebel) [inline csharp/ql/src/Linq/MissedFirstOrDefaultOpportunity.ql]: Perhaps, change the message slightly (then only `predicate` will have a "link").
```suggestion
  "This foreach loop returns the first sequence element satisfying a $@ - consider finding the element explicitly using '.FirstOrDefault(...)'.",
  is.getCondition(), "predicate"
```

File: csharp/ql/lib/Linq/Helpers.qll [hunk 3]
```diff
@@ -156,6 +181,32 @@ predicate missedWhereOpportunity(ForeachStmtGenericEnumerable fes, IfStmt is) {
   )
 }
 
+/**
+ * Holds if `foreach` statement `fes` could be converted to a `.FirstOrDefault()` call.
+ * That is, the loop contains a single `if` statement that accesses the loop variable,
+ * returns the loop variable when the condition matches, and is followed by a default return.
+ */
+predicate missedFirstOrDefaultOpportunity(
```
Comment (michaelnebel) [inline csharp/ql/lib/Linq/Helpers.qll]: Perhaps consider removing `ret` and `defaultRet` as predicate parameters as they are not needed as a part of the alert message.

File: csharp/ql/src/Linq/MissedFirstOrDefaultOpportunity.ql [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ *       maintainability
+ *       readability
+ *       language-features
+ */
+
+import csharp
+import Linq.Helpers
+
+from ForeachStmtGenericEnumerable fes, IfStmt is, ReturnStmt ret, ReturnStmt defaultRet
+where missedFirstOrDefaultOpportunity(fes, is, ret, defaultRet)
```
Comment (michaelnebel) [inline csharp/ql/src/Linq/MissedFirstOrDefaultOpportunity.ql]: `ret` and `defaultRet` are not used in the alert message. Consider using wildcards `_` or change the predicate to not include `ret` and `defaultRet` as parameters.

File: csharp/ql/src/Linq/MissedFirstOrDefaultOpportunityFix.cs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class MissedFirstOrDefaultOpportunityFix
+{
+    public static Operation FindOperation(System.Collections.Generic.IEnumerable<Operation> operations, string operationId)
```
Comment (michaelnebel) [inline csharp/ql/src/Linq/MissedFirstOrDefaultOpportunityFix.cs]: To align with the examples of the other `cs/linq/*`, perhaps remove `System.Collections.Generic` (optionally add a `using System.Collections.Generic` at the top of the file).

File: csharp/ql/lib/Linq/Helpers.qll [hunk 6]
```diff
@@ -20,6 +20,31 @@ private int numStmts(ForeachStmt fes) {
...
+    exists(DefaultValueExpr defaultValue |
+      defaultValue = ret.getExpr().stripCasts() and
+      (
+        defaultValue.getType() = elementType
+        or
+        hasNullDefault(elementType) and
+        hasNullDefault(defaultValue.getType())
+      )
+    )
+  )
```
Comment (michaelnebel) [inline csharp/ql/lib/Linq/Helpers.qll]: Note, that the size of this predicate (the number of tuples) can be very big as it only ties together the types of a return statement and a foreach statement (even tough they are in wildly different locations/methods). This can cause problems as CodeQL is evaluated bottom up.
The problem can be solved by adding the criteria, that the return statement should immediately follow the foreach statement (this is currently located in `missedFirstOrDefaultOpportunity`) .

```suggestion
   exists(BlockStmt enclosingBlock, int i, Type elementType |
    enclosingBlock.getStmt(i) = fes and
    enclosingBlock.getStmt(i + 1) = ret and
    elementType = fes.getVariable().getType()
  |
    ret.getExpr().stripCasts() instanceof NullLiteral and
    hasNullDefault(elementType)
    or
    exists(DefaultValueExpr defaultValue |
      defaultValue = ret.getExpr().stripCasts() and
      (
        defaultValue.getType() = elementType
        or
        hasNullDefault(elementType) and
        hasNullDefault(defaultValue.getType())
      )
    )
  )
```

**Review by michaelnebel — COMMENTED** [review]
Thank you very much! I was just thinking about the same thing, when I reviewed the other PR about the `missed-where` false positives 😄 .
I have added some initial comments.

File: csharp/ql/src/codeql-suites/csharp-security-and-quality.qls (hunk 1 above)
Comment (baywet) [reply csharp/ql/src/codeql-suites/csharp-security-and-quality.qls]: reverted in 6744d560243

File: csharp/ql/lib/Linq/Helpers.qll (hunk 3 above)
Comment (baywet) [reply csharp/ql/lib/Linq/Helpers.qll]: Thanks! Applied in c97f0f4b305

File: csharp/ql/src/Linq/MissedFirstOrDefaultOpportunityFix.cs (hunk 5 above)
Comment (baywet) [reply csharp/ql/src/Linq/MissedFirstOrDefaultOpportunityFix.cs]: Thanks! Applied in c97f0f4b305

File: csharp/ql/src/Linq/MissedFirstOrDefaultOpportunity.ql (hunk 4 above)
Comment (baywet) [reply csharp/ql/src/Linq/MissedFirstOrDefaultOpportunity.ql]: Thanks! Applied in c97f0f4b305

**Review by michaelnebel — COMMENTED** [review]
Thank you for addressing the comments!
It looks like the last commit rolls back some of the other changes.
I have opened a PR, which adds some commits on top of these great work you have been doing here. It is the last three commits of the PR seen [here](https://github.com/github/codeql/pull/22497). It should solve potential performance problems, update integration test expected output and add a change note.
Feel free to cherry-pick those comments (if you agree with the changes) - then we can run DCA (automated large scale testing) of the new query.

**Comment by baywet** [comment]
@michaelnebel thanks for putting these together, I've cherry picked all of them. b96bb465c19273f0a6881561e25c968fa868508d was creating conflicts but I think I got all the changes you intended to apply.

Let me know if you have any additional comments or questions.

File: csharp/ql/lib/Linq/Helpers.qll [hunk 7]
```diff
@@ -20,6 +20,36 @@ private int numStmts(ForeachStmt fes) {
...
+
+private predicate hasNullDefault(Type t) { t.isRefType() or t instanceof NullableType }
+
+private predicate returnsDefaultValueAfterForeach(ForeachStmt fes) {
+  exists(BlockStmt enclosingBlock, int i, Type elementType, ReturnStmt ret |
+    enclosingBlock.getStmt(i) = fes and
+    enclosingBlock.getStmt(i + 1) = ret and
+    elementType = fes.getVariable().getType()
+  |
+    ret.getExpr().stripCasts() instanceof NullLiteral and
```
Comment (michaelnebel) [inline csharp/ql/lib/Linq/Helpers.qll]: If we no longer want to flag cases where explicit casts are used (related to the last commit), then perhaps we should change the corresponding logic for the default value return as well.
```suggestion
    ret.getExpr().stripImplicit() instanceof NullLiteral and
```

File: csharp/ql/lib/Linq/Helpers.qll [hunk 8]
```diff
@@ -20,6 +20,36 @@ private int numStmts(ForeachStmt fes) {
...
+    hasNullDefault(elementType)
+    or
+    exists(DefaultValueExpr defaultValue |
+      defaultValue = ret.getExpr().stripCasts() and
```
Comment (michaelnebel) [inline csharp/ql/lib/Linq/Helpers.qll]:
```suggestion
      defaultValue = ret.getExpr().stripImplicit() and
```

File: csharp/ql/lib/Linq/Helpers.qll [hunk 9]
```diff
@@ -156,6 +186,26 @@ predicate missedWhereOpportunity(ForeachStmtGenericEnumerable fes, IfStmt is) {
...
+  // The loop only checks whether the current element is the first match.
+  is = firstStmt(fes) and
+  not exists(is.getElse()) and
+  numStmts(fes) = 1 and
+  // Condition relies on loop variable.
+  exists(VariableAccess va |
+    va.getTarget() = fes.getVariable() and
+    va = is.getCondition().getAChildExpr*()
+  ) and
+  not is.getCondition().getAChildExpr*() instanceof AwaitExpr and
```
Comment (michaelnebel) [inline csharp/ql/lib/Linq/Helpers.qll:225]: To address copilots comment
```suggestion
  not is.getCondition().getAChildExpr*() instanceof AwaitExpr and
  not fes.isAsync() and
```
However, I don't know how many FPs this will eliminate, as it will require that the enumerable that the `foreach` iterates over is both an `IEnumerable<T>` and `IAsyncEnumerable<T>`.

**Review by michaelnebel — COMMENTED** [review]
A few more comments 😄

File: csharp/ql/src/Linq/MissedFirstOrDefaultOpportunity.ql [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+ *       maintainability
+ *       readability
+ *       language-features
+ */
+
+import csharp
+import Linq.Helpers
+
+from ForeachStmtGenericEnumerable fes, IfStmt is
+where missedFirstOrDefaultOpportunity(fes, is)
```
[the remark this answers is not in the document]
Comment (baywet) [reply csharp/ql/src/Linq/MissedFirstOrDefaultOpportunity.ql:19]: @michaelnebel it seems that the changes you suggested re-activated this finding. What would be the best path forward here?

[the remark this answers is not in the document]
Comment (michaelnebel) [reply csharp/ql/src/Linq/MissedFirstOrDefaultOpportunity.ql:19]: Maybe it was co-incidental that this finding was re-activated (the AI findings are not deterministic). I just merged your other PR; Could you rebase this PR and push again?

[the remark this answers is not in the document]
Comment (baywet) [reply csharp/ql/src/Linq/MissedFirstOrDefaultOpportunity.ql:19]: merged. not rebased in case you were doing work locally.

**Review by michaelnebel — COMMENTED** [review]
@baywet : Thank you for the continued work on this PR 😄 
Some comments
- The indentation on the last change needs to be updated (failing formatting check).
- I tried running the new query against a 100 open source repositories and spot checked some of the alerts; None of these were related to lambda variable capture (they were just "plain vanilla" foreach statements followed by a `return null` statement). That is, I think we should just ignore such corner-cases.
- I also tried creating a new query, which is the conjunction of "missed where" and "missed first or default" (to investigate "high" warning that CoPilot issued yesterday) and ran it against 1000 open source repos - and there were no findings. I suspect that CoPilot didn't fully account for the changes you made to the "missed where" before issuing the warning.

If the CI is green, after the indentation issue is fixed, then I will trigger a DCA run to check the performance and results that the query produces. If everything looks acceptable, then we can probably merge.
Once again, thank you for your diligence!

**Comment by baywet** [comment]
@michaelnebel `codeql query format --in-place` makes no difference on the current rule, and I can't access the workflow run (I get a 404), do you have more details about the formatting issues?

File: csharp/ql/lib/Linq/Helpers.qll [hunk 11]
```diff
@@ -177,6 +207,28 @@ predicate missedWhereOpportunity(ForeachStmtGenericEnumerable fes, IfStmt is) {
...
+  not fes.isAsync() and
+  returnsLoopVariable(fes, is.getThen()) and
+fes.getElementType() = fes.getVariable().getType() and
```
Comment (michaelnebel) [inline csharp/ql/lib/Linq/Helpers.qll]: @baywet : This is most likely the location of the formatting issue (indentation)
```suggestion
  fes.getElementType() = fes.getVariable().getType() and
```

**Comment by baywet** [comment]
@michaelnebel Thanks for the suggestion, applied that and added a test case for the corner case you were mentioning I think.
I've left the where+firstOrDefault case you were mentioning out for now until we get additional data from running the DCA.

Let me know if you have any additional comments or questions.

**Review by michaelnebel — COMMENTED** [review]
Starting a DCA run.

File: csharp/ql/lib/Linq/Helpers.qll [hunk 12]
```diff
@@ -177,6 +207,29 @@ predicate missedWhereOpportunity(ForeachStmtGenericEnumerable fes, IfStmt is) {
...
+  not exists(is.getElse()) and
+  numStmts(fes) = 1 and
+  // Condition relies on loop variable.
+  exists(VariableAccess va |
+    va.getTarget() = fes.getVariable() and
+    va = is.getCondition().getAChildExpr*()
+  ) and
+  not is.getCondition().getAChildExpr*() instanceof AwaitExpr and
+  not fes.isAsync() and
+  not fes.getVariable().isCaptured() and
```
[the remark this answers is not in the document]
Comment (michaelnebel) [reply csharp/ql/lib/Linq/Helpers.qll:233]: This is a problem for all the LINQ queries as it is. I will open an issue for it, but it is out of scope to fix as a part of this PR.

**Review by michaelnebel — DISMISSED** [review]
DCA looks good.
- Performance appears to be unaffected.
- Including duplicates we found around 200 results in our DCA test suite (some projects are scanned both in traced and `build-mode: none`, which cause duplicate findings).
- I spot checked around thirty results and only found one result, where the re-write suggestion would capture an `in`/`out`/`ref` parameter (which is not allowed) - the remaining results looked good to me. The other LINQ queries suffer from the same problem, so I don't think it should block the merge of this PR.

An issue has been opened for addressing the FPs for LINQ queries, which suggests a re-write that would capture an `in`/`out` or `ref` parameter.

Thank you very much @baywet !

**Review by michaelnebel — COMMENTED** [review]
@baywet : Was just about to merge, but it turns out that a conflicting PR was merged this morning (it can be seen [here](https://github.com/github/codeql/pull/22492/). `ForeachStmt` has been renamed to `ForEachStmt` (and some similar renames).

Could you rebase this PR on main (or merge main into this branch) and update the naming? (otherwise we break the CI due to deprecation warnings).

**Comment by baywet** [comment]
@michaelnebel I've merged main into the PR and I think I've applied all the renames that were necessary?
Let me know if you have any additional comments or questions.

**Review by michaelnebel — COMMENTED** [review]
@baywet : Thank you!
I will also run DCA once more.

**Review by michaelnebel — APPROVED** [review]
LGTM!
